File: zlh/LeetCode/2023-11-21/501. 二叉搜索树中的众数.ts
function findMode(root: TreeNode | null): number[] {
  const res: number[] = [];
  const map: object = {};
  let maxCount: number = 0;
  let counts: number[] = [];

  const midTraversl = (root: TreeNode | null): null | undefined => {
    if (!root) return null;

    midTraversl(root.left);
    res.push(root.val);
    midTraversl(root.right);
  };

  midTraversl(root);

  for (let i: number = 0; i < res.length; i++) {
    let num = res[i];
    if (!map[num]) {
      map[num] = 1;
    } else {
      map[num] = map[num] + 1;
    }
  }

  for (let key in map) {
    if (map[key] > maxCount) {
      maxCount = map[key];
      counts = [Number(key)];
    } else if (map[key] === maxCount) {
      counts.push(Number(key));
    }
  }

  return counts;
}

File: zjj/501. 二叉搜索树中的众数.py
# Definition for a binary tree node.
from collections import Counter


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def findMode(self, root: TreeNode[TreeNode]) -> list[int]:
        node_vals = []
        ans = []

        def dfs(root):
            if not root:
                return
            node_vals.append(root.val)
            dfs(root.left)
            dfs(root.right)
        dfs(root)
        dic = Counter(node_vals)
        sort_dic = dict(
            sorted(dic.items(), key=lambda item: item[1], reverse=True))
        for key, val in sort_dic.items():
            if not ans:
                ans.append(key)
            else:
                if sort_dic[ans[0]] == val:
                    ans.append(key)
                else:
                    break
        return ans
